boltons/dictutils.py

```diff
@@ -693,4 +693,104 @@ def __reversed__(self):
             yield curr[KEY]
             curr = curr[PREV]
 
+
+class SentinelInt(int):
+    pass
+
+
+class FrozenDictBase(tuple):
+
+    def __getitem__(self, key):
+        if isinstance(key, SentinelInt):
+            return tuple.__getitem__(self, key)
+        else:
+            try:
+                return getattr(self, str(key.__hash__()))()
+            except AttributeError:
+                raise KeyError(key)
+            except IndexError:
+                raise KeyError(key)
+
+    def __iter__(self):
+        idx = 0
+        while True:
+            try:
+                yield self[SentinelInt(idx)][0]
+            except IndexError:
+                break
+            idx += 1
+
+    def items(self):
+        """Returns a list of tuples of data"""
+        return [(x, self[x]) for x in self]
```

Comment:
I know about the differences between py2 and 3 (as I pointed out). Since this library is suppose to be compatible with both I'm not sure what's the right path.
Your rationale waiting until design/approach is finalize makes perfect sense.

One thing you missed is my comment about the implementation of the items. My suggested implementation also creates a list, that's identical to what `items` produces, but is far more efficient.